### PR TITLE
Bug: Build with LibreSSL fails

### PR DESCRIPTION
**Pull request by minabarret**
- This codepath mimics some openssl-1.1 specific API and is enabled only
  for openssl 1.1 and higher versions. But, due to incompatible
  version numbering in libressl, it gets wrongly enabled with libressl
  versions that do not support the reqired API. As an easy workaround
  disable the feature when LIBRESSL_VERSION_NUMBER is defined.

Introduced With

https://github.com/OpenVPN/openvpn/commit/0e8a30c0b05c1e2b59a1dea0a6eab5daa1d9d9a1

The issue occurs when using libressl 2.6.4 with openvpn 2.4.5 using a modified openvpn-build environment

libressl:
 CPPFLAGS= ./configure --prefix=/var/tmp/bin/openvpn-build/generic/image/openvpn

openvpn:
./configure --prefix=/ --libdir=//lib --host= --target= --build= --program-prefix='' --enable-iproute2 --enable-pkcs11 --enable-lzo

error:

In file included from crypto_openssl.c:44:0:
openssl_compat.h:717:1: error: conflicting types for 'SSL_CTX_set_min_proto_version'
 SSL_CTX_set_min_proto_version(SSL_CTX *ctx, long tls_ver_min)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from openssl_compat.h:45:0,
                 from crypto_openssl.c:44:
/var/tmp/bin/openvpn-build/generic/image/openvpn/include/openssl/ssl.h:1175:5: note: previous declaration of 'SSL_CTX_set_min_proto_version' was here
 int SSL_CTX_set_min_proto_version(SSL_CTX *ctx, uint16_t version);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from crypto_openssl.c:44:0:
openssl_compat.h:746:1: error: conflicting types for 'SSL_CTX_set_max_proto_version'
 SSL_CTX_set_max_proto_version(SSL_CTX *ctx, long tls_ver_max)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from openssl_compat.h:45:0,
                 from crypto_openssl.c:44:
/var/tmp/bin/openvpn-build/generic/image/openvpn/include/openssl/ssl.h:1176:5: note: previous declaration of 'SSL_CTX_set_max_proto_version' was here
 int SSL_CTX_set_max_proto_version(SSL_CTX *ctx, uint16_t version);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Makefile:674: recipe for target 'crypto_openssl.o' failed
make[4]: *** [crypto_openssl.o] Error 1

Mina Barret does not use sourceforge or emails, but would be happy to use a unpatched openvpn with LibreSSL with the next release.

Thank you for your time.

# Thank you for your contribution

You are welcome to open PR, but they are used for discussion only. All
patches must eventually go to the openvpn-devel mailing list for review:

* https://lists.sourceforge.net/lists/listinfo/openvpn-devel

Please send your patch using [git-send-email](https://git-scm.com/docs/git-send-email). For example to send your latest commit to the list:

    $ git send-email --to=openvpn-devel@lists.sourceforge.net HEAD~1

For details, see these Wiki articles:

* https://community.openvpn.net/openvpn/wiki/DeveloperDocumentation
* https://community.openvpn.net/openvpn/wiki/Contributing

### CODE REVIEW

**Comment by cron2** [comment]
Hi,

On Sun, Mar 04, 2018 at 06:46:36AM +0000, minabarret wrote:
> - This codepath mimics some openssl-1.1 specific API and is enabled only
>   for openssl 1.1 and higher versions. But, due to incompatible
>   version numbering in libressl, it gets wrongly enabled with libressl
>   versions that do not support the reqired API. As an easy workaround
>   disable the feature when LIBRESSL_VERSION_NUMBER is defined.

See the release notes for 2.4.5.

If someone sends a patch, and feels responsible for maintaining LibreSSL
compatibility, we might consider it.

LibreSSL is not a supported crypto backend, and we are *not* going to test
every new LibreSSL version for "has it broken the API again?" - 2.4.5
has been tested on OpenBSD 6.0 with the LibreSSL version that comes with
that OS version and things work, so if LibreSSL is then breaking the API
again, the active developers very much do not care.

In other words: their approach to OpenSSL compatibility is wasting our
(precious and little) time.


Since you stated that you are not interested in following our normal
communication paths (mail) I am now closing the PR.

gert

-- 
now what should I write here...

Gert Doering - Munich, Germany                             gert@greenie.muc.de
